src/hex_api_key.erl

```diff
@@ -98,12 +99,36 @@ get(Config, Name) when is_map(Config) and is_binary(Name) ->
 %%     }}}
 %% '''
 %% @end
+
 -spec add(hex_core:config(), binary(), [permission()]) -> hex_api:response().
 add(Config, Name, Permissions) when is_map(Config) and is_binary(Name) and is_list(Permissions) ->
+    assert_valid_permissions(Permissions),
     Path = hex_api:build_organization_path(Config, ["keys"]),
     Params = #{<<"name">> => Name, <<"permissions">> => Permissions},
     hex_api:post(Config, Path, Params).
 
+assert_valid_permissions(Permissions) ->
+    FlatPerms = lists:flatmap(fun(P) -> maps:to_list(P) end, Permissions),
+    lists:foreach(fun({K,V}) -> assert_permission_key_value(K, V) end, FlatPerms).
+
+assert_permission_key_value(K, V) when not is_binary(K) andalso not is_binary(V) ->
+     Format = <<"expected permission key and value to be binaries, got ~p => ~p">>,
```

Comment:
I like that this error message talks about both key and value. I think we could have it instead of more specific ones which would simplify the code slightly. In fact, wdyt about the following?

```
expected permission to be a map of #{<<"domain">> => binary(), <<"resource">> => binary()}, got: #{bad => bad, worse => worse}
```

Reply:
That makes sense to me 👍

Reply:
yes, I'd use orelse.

I believe we'd always have just two k/v pairs so I think dumping the whole map is fine.

Reply:
@wojtekmach If we dump the whole map, we'll end up with line wrapping, are we ok with that? 

Example : 

```
1> Res =  io_lib:format(<<"expected permission to be map of #{<<\"domain\">> => binary(), <<\"resource\">> => binary()}, got : ~p">>, [#{domain => eh, resource => eh}]).
[101,120,112,101,99,116,101,100,32,112,101,114,109,105,115,
 115,105,111,110,32,116,111,32,98,101,32,109,97,112|...]
2> iolist_to_binary(Res).
<<"expected permission to be map of #{<<\"domain\">> => binary(), <<\"resource\">> => binary()}, got : #{domain =>\n        "...>>
```

The `...` is of course just from doing that in the shell. I'm ok with line wrapping personally since it'll just end up going to stdout in the case of rebar3_hex, but let me know if you're good with that. If not, then we can opt not to dump the full map and instead do it on k/v basis like below and avoid line wrapping: 

```
1> Res =  io_lib:format(<<"expected permission to be map of #{<<\"domain\">> => binary(), <<\"resource\">> => binary()}, got : #{~p => ~p}">>, [domain, eh]).
[101,120,112,101,99,116,101,100,32,112,101,114,109,105,115,
 115,105,111,110,32,116,111,32,98,101,32,109,97,112|...]
2> iolist_to_binary(Res).                                                                                                                                 
 <<"expected permission to be map of #{<<\"domain\">> => binary(), <<\"resource\">> => binary()}, got : #{domain => eh}">>
3>
```

Or like the above but ... 

```
<<"expected permission domain key/value to be #{<<\"domain\">> => binary()}, got : #{domain => eh}">>
```

Likewise if we encounter that for resource key/value : 

```
<<"expected permission resource key/value to be #{<<\"resource\">> => binary()}, got : #{resource => eh}">>
```

Yet another option would be to raise with something like `{error, {invalid_permissions, #{domain => eh, resource => eh}}}` and provide a format_error/1 function like we do on `src/hex_tarball.erl`, but raising doesn't feel in that case.

But yeah, just let me know.

Reply:
> If we dump the whole map, we'll end up with line wrapping, are we ok with that?

Good catch, that'd be not ideal. How about this?

```erlang
> Map = #{domain => eh,resource => eh}.
#{domain => eh,resource => eh}
> iolist_to_binary(["expected permissions to be a map with binary keys and values, got: ", io_lib:format("~p", [Map])]).
<<"expected permissions to be a map with binary keys and values, got: #{domain => eh,resource => eh}">>
```

> The ... is of course just from doing that in the shell.

Yeah that's not ideal either. But that's just in the shell right, e.g. in error reports we'd get the whole binary? If so I'm fine with that, if the user is doing things in the shell, they have access to the map so they don't need it printed fully for them. We can cut the length slightly like this:

```diff
- expected permissions to be a map with binary keys and values, got: 
+ expected permissions to have binary keys and values, got: 
```

If we return a string instead of a binary, the shell won't wrap:

```erlang
1> Map2 = #{domain => eh,resource => eh, another_key => lorem_ipsum_dolor_sit_amen}.

2> iolist_to_binary(["expected permissions to be a map with binary keys and values, got: ", io_lib:format("~p", [Map2])]).
<<"expected permissions to be a map with binary keys and values, got: #{another_key => lorem_ipsum_dolor_sit_amen,domai"...>>

3> binary_to_list(iolist_to_binary(["expected permissions to be a map with binary keys and values, got: ", io_lib:format("~p", [Map2])])).
"expected permissions to be a map with binary keys and values, got: #{another_key => lorem_ipsum_dolor_sit_amen,domain => eh,resource => eh}"
```

Is this enough of a reason to return errors as strings? I think it might be. You mentioned `format_error/1`, which is a good idea too, that function usually returns strings too? Always strings? If so, I'd go with returning strings then.

Otherwise, returning an error tuple and providing format_error/1 seems like the best option but it just feels like an overkill in this particular case, I'm not sure I want to have the format_error/1 for every single hex_api_* module.

>  If not, then we can opt not to dump the full map and instead do it on k/v basis like below and avoid line wrapping:

I'd go with the whole map but it is not a deal breaker for me.

Reply:
> > Map = #{domain => eh,resource => eh}.
#{domain => eh,resource => eh}
> iolist_to_binary(["expected permissions to be a map with binary keys and values, got: ", io_lib:format("~p", [Map])]).
<<"expected permissions to be a map with binary keys and values, got: #{domain => eh,resource => eh}">>

☝️ that'll work.